src/app/clusters/dishwasher-alarm-server/dishwasher-alarm-server.cpp

```diff
@@ -0,0 +1,458 @@
+    {
+        return Status::Failure;
+    }
+    return Status::Success;
+}
+
+static Status ResetHandler(const app::ConcreteCommandPath & commandPath, const BitMask<AlarmMap> alarms)
+{
+    EndpointId endpoint = commandPath.mEndpointId;
+
+    if (!DishwasherAlarmServer::Instance().HasResetFeature(endpoint))
+    {
+        ChipLogProgress(Zcl, "Dishwasher Alarm feature: Unsupport Reset Command");
+        return app::ToInteractionModelStatus(EMBER_ZCL_STATUS_UNSUPPORTED_COMMAND);
+    }
+
+    // A server that is unable to reset alarms SHALL respond with a status code of FAILURE
+    Delegate * delegate = DishwasherAlarm::GetDelegate(endpoint);
+    if (delegate && !(delegate->ResetAlarmsCallback(alarms)))
+    {
+        ChipLogProgress(Zcl, "Unable to reset alarms");
+        return Status::Failure;
+    }
+
+    // The cluster will do this update if delegate.ResetAlarmsCallback() returns true.
+    if (DishwasherAlarmServer::Instance().ResetStateValue(endpoint, alarms) != EMBER_ZCL_STATUS_SUCCESS)
+    {
+        ChipLogProgress(Zcl, "reset alarms fail");
+        return Status::Failure;
+    }
+    return Status::Success;
+}
+
+bool emberAfDishwasherAlarmClusterResetCallback(app::CommandHandler * commandObj, const app::ConcreteCommandPath & commandPath,
+                                                const Commands::Reset::DecodableType & commandData)
+{
+    auto & alarms = commandData.alarms;
+
+    Status status = ResetHandler(commandPath, alarms);
+    commandObj->AddStatus(commandPath, status);
+
+    return true;
+}
+
+bool emberAfDishwasherAlarmClusterModifyEnabledAlarmsCallback(app::CommandHandler * commandObj,
+                                                              const app::ConcreteCommandPath & commandPath,
+                                                              const Commands::ModifyEnabledAlarms::DecodableType & commandData)
+{
+    auto & mask   = commandData.mask;
+    Status status = modifyEnabledHandler(commandPath, mask);
+    commandObj->AddStatus(commandPath, status);
+
+    return true;
+}
+
+/**
+ * If the Dishwasher Alarm Cluster has Latch attribute, should call this function to handle the
+ * State attribute.
+ * If each bit set in the Latch attribute is true, the corresponding bit set in the
+ * State attribute can change from false to true, but unable to change from true to false;
+ * If each bit set in the Latch attribute is flase, the corresponding bit set in the
+ * State attribute can change to true or false.
+ * @param[in] latch The current latch attribute.
+ * @param[in] currentState The current State attribute.
+ * @param[in] newState The State attribute which want to set.
+ * @return The State attribute which want to set.
+ */
+static BitMask<AlarmMap> SetStateByLatch(const BitMask<AlarmMap> latch, const BitMask<AlarmMap> currentState, const BitMask<AlarmMap> newState)
+{
+    uint32_t rawLatch = latch.Raw();
```

Comment:
Why are we going to Raw() here?

But more to the point, I know the spec says "requires an explicit reset using the Reset command", but does that really mean not possible to reset via explicit physical interaction with the device?  @marktrayer As in, I think this function should not exist at all; if the application running on the device wants to set the state, we should let it set the state.

Reply:
@rbultman can chime in as well.  But latched in the most straightforward sense means that the alarm isn't cleared when the alarm condition clears.  That is, it remains active until some explicit action takes place to clear it.  This could be a RESET() command, it could be (depending on the device) some explicit action that the app is handling (e.g. "press and hold the 'lock' button for 3 seconds").

Reply:
Right.  That makes sense to me, and would mean this function should not exist: the app itself should be able to clear latched alarms if it wants to.

Reply:
From the test case of dishwasher alarm, if the Latch bit is set ,  the bit of State  can't reset. I suggest to implement this logic in SDK rather than in application.
https://github.com/CHIP-Specifications/chip-test-plans/blob/master/src/cluster/alarm_testcase.adoc

So this function is use the judge the State attribute while the Latch bit is set. So do you have suggestion to fix the issue? @bzbarsky-apple

Reply:
@marktrayer's comment is correct.  When a latchable alarm condition occurs, that alarm status bit should be 1 until both the alarm condition has cleared and the reset command has been sent with the appropriate mask.

To set a latchable alarm:
state = alarm condition AND supported AND mask

The alarm bit in state shall remain 1 until:
state = !alarm condition AND reset command

Reply:
> From the test case of dishwasher alarm, if the Latch bit is set , the bit of State can't be reset

Which exact part of which exact test plan are you looking at, @fengdx4 ?  

When I look at the dishwasher alarm test plans, the only mentions of "latch" are:

* In TC-DISHALM-2.1 step 3, check that the Latch attribute is a 32-bit integer.
* In TC-DISHALM-3.1 step 2d, checks that the alarm was not cleared if the Latch bit is set.  That said, step 2c is _extremely_ unclear, since it says to actually change the State attribute, not get rid of the alarm condition.  This part seems like a bug in the test plan...  What it should be saying is get rid of the alarm condition, and then observe how State evolves.  If it says that State changed, as now, then ***State must change***.
* TC-DISHALM-3.2 step 2d, same as above.
* Same for the other test plans, alarm by alarm.

Reply:
The test plan definitely needs work.

Reply:
@bzbarsky-apple  I totally understood.   The test plan operate only one bit in one time, it's workable for the application call the api to get its internal concept of "state" will not change when the alarm condition goes away if latch bit set to true.  
In this case TC-DISHALM-3.1  , application may call the api in several steps.
Step 2a.
 //1. call  `state = GetLatchValue(xx)`  to get the Latch State, 
// 2. call `state.GetField(xx, AlarmMap::kInflowError)` to the the corrsponding bit Latch for AlarmMap::kInflowError
// 3. application may write this if-else judge: (just like Reset-Set-flip-flop)
if  LatchBit for AlarmMap::kInflowError is true,  it need to jduge the corresponding bit of current State  and newState (this also need call `GetField` to get corresponding bit`)
            if corresponding bit of current state is  false
                     then newState can override  corrsponding bit of the current State to true. (This need call `SetField` or clear corresponding bit)
            else
                     keep the currentState. (Also need call `SetField`)
else if  LatchBit for AlarmMap::kInflowError is false 
           use corresponding bit of newState to override the corresponding bit of currentState
            
This is code for only one Test step, so for other test step, we also need to repeat this code. 

What's more, in practical applications( MCU+Wi-Fi appliance Dishwasher) , MCU provide the AlarmMap data `State` to Wi-Fi module together) , the application need to set each bit of AlarmMap for `State`.  So if we can call `SetStateValue() ` of SDK in one time, no need to care the Latch state in application, it may make more sense.

  =====
   1.  Init
    //Support 10 1111  (47) (0x2F)
    //Mask     10 1111  (47) (0x2F)
    //Latch     00 0011  (3)  (0x03)
    //State      00 0111  (7)  (0x07)   //Mcu report State together
    //FeatureMap 1
====
   2. MCU report the State
 //State      00 0010  (2)  (0x02)   //Mcu report State together
  3. SDK handle and return the State to Cluster Client ,
  //State     00 0011  (3)  (0x03)  

Is it make sense? Or this logic need to put in application may be more better? Do we need a quick talk? @rbultman

Reply:
I made some changes to remove the function @bzbarsky-apple mentioned.  As it stands, the device must call ResetStateValue in order to reset the latched bits.  I'll start working on the test plans now.

Reply:
@mideayanghui I would think the application has the "state with all the latched bits still set, even if the underlying condition is gone" somewhere, because it's got lights and whatnot showing that state, no?  And possibly audible alarms and so on....

But maybe I just don't understand how the applications using this cluster are intended to be structured?